Introducing support for Netatmo Welcome Camera

jabesq opened the pull request:
**Description:**
This change is introducing two things: 
- Global component for Netatmo devices: it allows to share credentials between Netatmo Weather Station, Welcome camera and also for Thermosat (even if not supported yet)
- Support for Netatmo Welcome Camera: Currently it only provides the live feed of the camera. The final goal will be to also to use this camera as a binary sensor and a device tracker 

**Related issue (if applicable):** fixes #

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/542

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
#Netatmo Component
netatmo:
    api_key: API_KEY
    secret_key: SECRET_KEY
    username: username
    password: password

#Cameras
camera:
    platform: netatmo
    home: home_name (Optional)
    cameras: (Optional)
        - camera_name1
        - camera_name2

sensor:
  platform: netatmo
  modules:
    module_name1:
      - temperature
      - humidity
      - noise
      - pressure
      - co2
      - rain
      - sum_rain_1
      - sum_rain_24
    module_name2:
      - temperature
    rainmeter_name3:
      - rain
      - sum_rain_1
      - sum_rain_24
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If code communicates with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [x] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
- [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] Tests have been added to verify that the new code works.

jabesq committed:
Introducing the Netatmo component

As Netatmo is providing several type of device (sensor, camera), a new Netatmo
component needs to be created in order to centralize the Netatmo login data.
Currently this change only impacts the Netatmo Weather station

jabesq committed:
Add new Netatmo library

This new API will provide access to the Welcome Camera

jabesq committed:
Basic support for Netatmo Welcome camera

This change introduces support for Netatmo Welcome camera. Currently, it will
add all detected camera to Home Assistant, camera filtering (similar to the one
used for weather station modules) will be added later

jabesq committed:
Remove useless REQUIREMENTS

jabesq committed:
Fixes for Netatmo Welcome support

jabesq committed:
Allow to filter Welcome cameras by name and/or home

jabesq committed:
Update requirements for Netatmo components

jabesq committed:
Fix multi-camera support for Welcome

jabesq committed:
Fix pep8 error/warning

rmkraus commented:
Is it possible to have the Netatmo component search for the cameras and sensors then create what is found? Take a look at the Hue component, it does this.

jabesq commented:
It already works like this for the camera (but I detail on the documentation how to filter for a specific camera if the user has several and want to display only one)
For the sensor, I just kept the behavior of the current netatmo sensor. The only change is that credentials are now stored inside a new global component Netatmo

rmkraus commented:
Great! 👍  

Maybe a later PR could cover discovery of Netatmo sensors, but that would definitely be too much for one PR.

rmkraus commented on homeassistant/components/camera/netatmo.py:
```diff
@@ -0,0 +1,98 @@
+"""
+Support for the NetAtmo Weather Service.
```
This looks like it needs a quick update 😉

rmkraus commented on homeassistant/components/netatmo.py:
```diff
@@ -0,0 +1,61 @@
+"""
+Support for the NetAtmo Weather Service.
```
Another one... sorry to nitpick.

rmkraus commented on homeassistant/components/sensor/netatmo.py:
```diff
@@ -43,29 +36,9 @@
-    data = NetAtmoData(authorization, config.get(CONF_STATION, None))
+    """Setup the available NetAtmo weather sensors."""
+    netatmo = get_component('netatmo')
+    data = NetAtmoData(netatmo.NETATMO_AUTH, config.get(CONF_STATION, None))
```
It actually looks like sensor discovery is triggered automatically. The discovery looks like it will stop here, however, as discovery won't provide config. The name of the station could be provided in the discovery_info to solve this. Maybe at a future time.

rmkraus replied:
Just checking here, what happens if None is put into NetAtmoData as the station name? Also, what will happen here if bad authorization details are provided?

jabesq replied:
This parameter was newly added to the Netatmo sensor it allows to limit the sensors to a specific Weather Station, but if None is provided then all user's weather stations will be parsed. I did not work on the PR that was introducing this change, but I confirm that it did not break a thing

rmkraus commented on homeassistant/components/camera/netatmo.py:
```diff
@@ -0,0 +1,98 @@
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """Setup access to Netatmo Welcome cameras."""
+    netatmo = get_component('netatmo')
+    home = config.get(CONF_HOME, None)
```
This config option doesn't seem to be documented in your PR.

jabesq replied:
I just add it, it is an optional config

rmkraus commented on homeassistant/components/camera/netatmo.py:
```diff
@@ -0,0 +1,98 @@
+    """Setup access to Netatmo Welcome cameras."""
+    netatmo = get_component('netatmo')
+    home = config.get(CONF_HOME, None)
+    data = WelcomeData(netatmo.NETATMO_AUTH, home)
```
What happens here it NETATMO_AUTH is bad?

jabesq replied:
if NETATMO_AUTH is bad then the netatmo component fails to load and then home-assistant can't load the camera.netatmo  component (as their is dependency) 
error message: homeassistant.bootstrap: Unable to prepare setup for platform camera.netatmo because dependency netatmo could not be initialized

rmkraus commented:
Looks really good. I had a few nitpicky docstring comments (sorry) and a few more questions.

In particular, how will the camera and sensor platforms react when improper credentials are provided? They don't seem to individually check to ensure that these credentials are good. Especially since this would be a breaking change, the platforms should probably put good information in the logs.

Now that I'm typing, what if the sensors platform still accepted the credentials, as is it did before, but threw a deprecation warning to the log? That all helps ease the transition of breaking changes.

jabesq committed:
This commit also adds improved logging for bad credentials

jabesq commented on homeassistant/components/camera/netatmo.py:
```diff
@@ -0,0 +1,100 @@
+        add_devices_callback([WelcomeCamera(data, camera_name, home)])
+
+
+class WelcomeCamera(Camera):
```
I don't get it what you want here @balloobbot

balloob commented on homeassistant/components/camera/netatmo.py:
```diff
@@ -0,0 +1,98 @@
+                self.camera_names.append(camera['name'])
+        return self.camera_names
+
+    def update(self):
```
You should throttle this call, otherwise it will be called once for each camera that is failing inside `camera_image`.

balloob commented on homeassistant/components/sensor/netatmo.py:
```diff
@@ -32,40 +36,54 @@
+                        "Netatmo hub component, See: "
+                        "https://home-assistant.io/components/netatmo/")
+
+        # All this should be removed once the deprecated method will not be
```
Let's remove the deprecated method and I will add it as a breaking change.

jabesq committed:
Add Throttle decorator for Welcome update function

As the update function updates the data for all cameras, we should prevent this
function to be called several time during an interval

balloob commented:
Looks good! :dolphin:

balloob commented:
@jabesq it looks like you deleted the 0.5 release of the Python Netatmo lib that you had Home Assistant point at for this PR, breaking Home Assistant. What's going on ?

jabesq commented:
Ooops, I did a force rebase on philipelt's netatmo API (which has credentials to the pypi archive) as I plan to retake my work for the Welcome camera but my rebase might have delete my tag, I'll recreate it so home-assistant can find it again. But my goal was that home-assistant to use directly the pypi archive.
